README.md
# What is it?

6502.ts strives to be a collection of emulators for 6502 based systems written in Typescript.
It currently provides

 * A cycle-exact [CPU emulator](doc/cpu.md) and a generic [debugging environment](doc/vanilla_debugger.md)
   any attached hardware.
 * A minimal [hardware monitor](doc/ehbasic_monitor.md) for executing the
   [EhBasic interpreter](https://github.com/jefftranter/6502/tree/master/asm/ehbasic)
   and Klaus Dormann's
   [6502 testsuite](https://github.com/Klaus2m5/6502_65C02_functional_tests)
 * A full-featured [Atari 2600 emulator](doc/internals/stella.md), including sound, CRT phosphor simulation,
   debugging environment and a more user oriented ["stellerator" frontend](doc/stellerator.md).

The projects targets both NodeJS and the browser as runtime environments (no video or audio on node,
of course).

# Atari 2600 emulation

6502.ts implements a full-fledged emulator for the Atari 2600 VCS. Apart from the
debugger, there are two ways to use the emulator

 * **Stellerator** allows to import, manage and play VCS ROMs in the browser. ROMs are
   stored locally in the browser. It is responsive and can be used offline as a homescreen
   application on iOS. Stellerator is hosted on [github.io](https://6502ts.github.io/stellerator-ng).
 * **Stellerator (legacy)** is the old version of Stellerator. It is not developed anymore, but
   still hosted on [github.io](https://6502ts.github.io/stellerator).
 * **Stellerator embedded** offers a rich API to embed VCS emulation into web sites
   and web applications. Please check out the
   [documentation](https://6502ts.github.io/typedoc/stellerator-embedded/)
   for more information.

# Tooling

The emulation core is written in [Typescript](https://www.typescriptlang.org).
The Stellerator frontend is mostly written in the [Elm](https://elm-lang.org) and uses
a custom DOS-style CSS theme written in [SASS](https://sass-lang.com). Builds are
created with [webpack](https://webpack.js.org), and [Yarn](https://yarnpkg.com/lang/en/)
is used for package management.

# Building and development

6502.ts uses [Yarn](https://yarnpkg.com/lang/en/) for package management, so you'll
have to install it first. After checking out the repos, do

```
    yarn install
```

to pull in the required packages.

## Tests

Test are run with

```
    yarn test
```

## Stellerator

The Stellerator frontend is built by doing

```
    yarn build
```

After the build has finished, the compiled frontend can be found in the `dist` directory.

For development, a dev server on http://localhost:6502 in watch mode can be fired up with

```
    yarn start
```

Development builds are written to `dist-dev`. For these builds, Elm runs in development mode,
the compiled code is unminified, and the service worker registration is skipped.


## CLI

Several tools are available as CLI only (notably the ehBasic hardware monitor). Those can be
built by doing

```
    yarn tsc
```

After building, the compiled (modularized) code is available in the `compiled` directory,
and the tools are availabe in `compiled/bin`.

# NPM package

The core parts of the emulator are available as `6502.ts` on
[NPM](https://www.npmjs.com).
The package includes TypeScript typings and can be used directly in TypeScript projects.

Most APIs are pretty stable by now, but not documented yet, with the exception
of Stellerator embedded. Please check out the
[documentation](https://6502ts.github.io/typedoc/stellerator-embedded/)
of Stellerator embedded for more details.

# License and credits

## The MIT license

Copyright (c) 2014 -- 2020 Christian Speckner and contributors

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

## Third party code

These license conditions do not apply to the contents of the `aux` directory which
mostly were written by folks other than me. Specifically:

 * `ehBasic` was taken from Jeff Tranters repository
   [here](https://github.com/jefftranter/6502/tree/master/asm/ehbasic)
 * Klaus Dormann's 6502 testsuite (found in `aux/6502_suite`) was taken from his
   repository on github [here](https://github.com/Klaus2m5/6502_65C02_functional_tests)
   and is licensed under the GPL.
 * The `red_line` sample was taken from Kirk Israel's 2600 programming tutorial on
   [AtariAge](http://www.atariage.com/2600/programming/2600_101/03first.html). The
   `line_test` is a modified version of this code.
 * The `playfield_1` sample was taken from Kirk Israel's [2600 cookbook](http://alienbill.com/2600/cookbook/playfield.html).
 * `playfield_2` was taken from Andrew Davie's 2600 programming excercises on
   [AtariAge](http://atariage.com/forums/topic/28219-session-15-playfield-continued/).
 * The `macro.h` and `vcs.h` headers can be found free-floating on the internet
 * `flapping` is a homebrew game written by Kirk Israel available
   [here](http://alienbill.com/2600/flapping/).

2600 cartridge type detection code (and some of the more exotic cartridge types)
were modelled after the excellent [Stella](https://stella-emu.github.io)
emulator. Initial 2600 audio code was contributed by Martin Schröder. ARM
support in DPC+ is powered by David Welch's
[thumbulator](https://github.com/dwelch67/thumbulator), transpiled to JS with
[emscripten](http://kripken.github.io/emscripten-site/). Cycle-accurate PCM
audio is derives from work done by Chris Brenner. The favicon and homescreen icons
in Stellerator were taken from Stella.


CHANGELOG.md
# 1.1.4

## All

* Another supercharger BIOS fix -> fixes splash screen on Communist Mutants from
  Space.

# 1.1.3

## All

 * Fix handling of ZP memory in supercharge BIOS -> fixes loads in Survival Island
 * Fix transition counting in supercharger emulation -> fixes bad opcode in
   Survival Island

# 1.1.2

## stellerator-embedded

 * Fix filenames in documentation.

# 1.1.1

## stellerator-embedded

 * Fix crash if `enableKeyboard` is disabled and `resetViaKeyboard` is
   enabled.

# 1.1.0

## All

 * Rewrite build process with webpack instead of rollup.

## stellerator-embedded

 * Switch stellerator-embedded to svideo by default in order to avoid
   graphical glitches on iOS with composite.

# 1.0.6

## All

 * Better version of my 1.0.5 video fix.

# 1.0.5

## All

 * Fix Moiré artifacts in composite effect on Apple silicon.

# 1.0.4

## stellerator-embedded

 * Fix touch I/O


# 1.0.3

## stellerator-embedded

 * First release after the 1.x rewrite
 * Support the new video settings
 * Rework the documentation
 * Example page: load the ROM via `fetch`

## Stellerator

 * Documentation

# 1.0.2

## VcsRunner

 * Support for running to RTS
 * Conditional traps

# 1.0.1

## All

 * Move dasm.js to deps

# 1.0.0

## All

 * Add VcsRunner for unit testing VCS software

# 1.0.0-beta11

## All

 * Fix an initialization issue in Missile emulation; fixes BMX Airmaster.

## Stellerator

 * Fix fullscreen toggle via touch on iOS 13.4. Unfortunately, this means that the "alt"
   quadrant of the canvas can't be used as a handle for scrolling anymore.

## Stellerator-embedded

 * Defunct, still needs plumbing for the new video


# 1.0.0-beta10

## Stellerator

 * Don't nuke stored data if an exception occurs during the initial load

## Stellerator-embedded

 * Defunct, still needs plumbing for the new video

# 1.0.0-beta9

## Stellerator

 * New WebGL-only video infrastructure with GPU based postprocessing
 * Phosphor simnualation modelled after Stella
 * TV effects  based on a libretro shader by Themaister
 * Scanlines
 * QIS scaling
 * More stable framerates

## Stellerator-embedded

 * Defunct, still needs plumbing for the new video

# 1.0.0-beta9

## Stellerator

 * New WebGL-only video infrastructure with GPU based postprocessing
 * Phosphor simnualation modelled after Stella
 * TV effects  based on a libretro shader by Themaister
 * Scanlines
 * QIS scaling
 * More stable framerates

## Stellerator-embedded

 * Still needs plumbing for the new video

# 1.0.0-beta.8

## Stellerator

 * Re-enable gamepad supports

## Stellerator-embedded

 * Fix gamepad mapping for select

# 1.0.0-beta.7

## Stellerator

 * Re-add the possibility to save ROMs (currently defunct on iOS)
 * Performance improvements in ARM games

## Stellerator-embedded

 * Allow to run without canvas
 * Allow to change canvas element at runtime
 * Add data tap for sending data from 6507 code to the host

# 1.0.0-beta.6

## Stellerator

 * Move version, changelog and license to dedicated "About" page
 * Don't show update notice on first start

# 1.0.0-beta.5

## Stellerator

 * Styling fixes on help page.

# 1.0.0-beta.4

## Stellerator

 * Release keyboard and mouse on leaving emulation view -> fixes nonfunctional
   keys after starting emulation
 * Add changelog to help page
 * Offer changelog after update

# 1.0.0-beta.3

## Stellerator

 * Fix keyboard navigation order

# 1.0.0-beta.2

## Stellerator

 * Minor bug fixes, UI improvements and optimizations in Stellerator
 * Reenable forgotten paddle driver. Ups. Paddles work again

## Stellerator-embedded

 * Reenable stellerator-embedded in build system

## All

 * Change license back to MIT
 * Remove "curtain" screen when loading supercharger loads in order to steer
   100% clear of potential GPL vs. MIT issues

# 1.0.0-beta.1

## Stellerator

 * Stellerator frontend rewritten in Elm
 * Replace deprecated cache manifest for Stellerator with service worker

## All

 * Build system replaced with yarn scripts + Rollup
 * Support for the Pink Panther prototype banking scheme
 * Workaround for startup issues in Activison Decathlon
 * Stellerator-embedded and debug frontends temporarily removed from build
 * Travis deployment temporarily disabled

# 0.9.14

## All

 * Support for CDFJ banking
 * CommaVideo cartridge support

# 0.9.13

## All

 * New CPU core with one bus access per cycle
 * Support 2 / 4 / ... / 1024 byte ROMs
 * Audio fixes for Safari iOS / MacOS: audio now reliably starts after the first interaction on both platforms

# 0.9.12

## Stellerator embedded

 * Fix pause via touch controls

# 0.9.11

## Stellerator embedded

* include support for touch controls

## All

 * Update dependencies

# 0.9.10

## All

 * Fix CDF music mode.

# 0.9.9

## All

 * Improvements to playfield collisions during HBLANK. Fixes a glitch in
   Thrust.

# 0.9.8

## Stellerator

 * UI fixes for mobile devices

## Stellerator embedded

 * Reduce default volume to 70%.

## All

 * Work around audio issues on iOS
 * Implement a poor man's fullscreen mode for iOS devices
 * Restrict HiDPI scaling in order to avoid slowdowns in Chrome

# 0.9.7

## All

 * Fix an extreme edge case of late HMOVE followed by another HMOVE. This
   fixes a spurious line at the left side of the screen in panda chase.

# 0.9.6

## All


Gruntfile.js
/*
 *   This file is part of 6502.ts, an emulator for 6502 based systems built
 *   in Typescript.
 *
 *   Copyright (C) 2014 - 2018 Christian Speckner & contributors
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License along
 *   with this program; if not, write to the Free Software Foundation, Inc.,
 *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

const path = require('path'),
    envify = require('envify/custom'),
    cp = require('child_process');

const TSLINT_OPTIONS = `--project ./tsconfig.json --config ./tslint.json --exclude '**/*.js'`;

const LICENSE_PREAMPLE = `
/*
 *   This file is part of 6502.ts, an emulator for 6502 based systems built
 *   in Typescript.
 *
 *   Copyright (C) 2014 - 2018 Christian Speckner & contributors
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   Please go to the project's github page at
 *
 *   https://github.com/6502ts/6502.ts
 *
 *   for more details and for a copy of the GNU General Public License v2.
 */
`;

module.exports = function(grunt) {
    'use strict';

    grunt.loadNpmTasks('grunt-ts');
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-browserify');
    grunt.loadNpmTasks('grunt-mocha-test');
    grunt.loadNpmTasks('grunt-http-server');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-notify');
    grunt.loadNpmTasks('grunt-template');
    grunt.loadNpmTasks('grunt-exorcise');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-postcss');
    grunt.loadNpmTasks('grunt-contrib-copy');
    grunt.loadNpmTasks('grunt-exec');
    grunt.loadNpmTasks('grunt-typedoc');
    grunt.loadNpmTasks('grunt-contrib-compress');

    grunt.loadTasks('./grunt');

    let buildId;
    try {
        buildId = cp
            .execFileSync('git', ['rev-parse', '--short=6', 'HEAD'], {
                encoding: 'utf-8'
            })
            .replace(/\s/, '');
    } catch (e) {
        buildId = '[unknown]';
    }

    grunt.initConfig({
        ts: {
            main: {
                tsconfig: {
                    tsconfig: '.',
                    passThrough: true
                }
            }
        },

        clean: {
            base: [
                '.tscache',
                'web/js/compiled',
                'web/js/installed',
                'tests/fixtures/fs_provider/blob.json',
                'web/stellerator.html',
                'build',
                'compiled',
                'typedoc'
            ],
            mrproper: ['web/js/installed', 'web/css/installed']
        },

        exec: {
            tslint: {
                cmd: `${path.join(__dirname, 'node_modules', '.bin', 'tslint')} ${TSLINT_OPTIONS}`
            }
        },

        browserify: {
            options: {
                configure: b => b.plugin('tsify', { project: __dirname }),
                browserifyOptions: {
                    debug: true
                }
            },
            ehBasicCLI: {
                dest: 'web/js/compiled/ehBasicCLI.js',
                src: [],
                options: {
                    alias: {
                        ehBasicCLI: './src/web/ehBasicCLI.ts'
                    }
                }
            },
            debuggerCLI: {
                dest: 'web/js/compiled/debuggerCLI.js',
                src: [],
                options: {
                    alias: {
                        debuggerCLI: './src/web/debuggerCLI.ts'
                    }
                }
            },
            stellaCLI: {
                dest: 'web/js/compiled/stellaCLI.js',
                src: [],
                options: {
                    alias: {
                        stellaCLI: './src/web/stellaCLI.ts'
                    },
                    cacheFile: 'foo.json'
                }
            },
            stellerator_worker: {
                options: {
                    configure: b => b.plugin('tsify', { project: path.join(__dirname, 'worker') })
                },
                dest: 'web/js/compiled/worker/stellerator_worker.js',
                src: ['worker/src/main/stellerator.ts']
            },
            video_pipeline_worker: {
                options: {
                    configure: b => b.plugin('tsify', { project: path.join(__dirname, 'worker') })
                },
                dest: 'web/js/compiled/worker/video_pipeline_worker.js',
                src: ['worker/src/main/video-pipeline.ts']
            },
            stellerator_dev: {
                dest: 'web/js/compiled/stellerator.js',
                src: ['src/web/stella/stellerator/main.tsx'],
                options: {
                    configure: b =>
                        b.plugin('tsify', { project: __dirname }).transform(
                            envify({
                                NODE_ENV: 'development',
                                STELLERATOR_WORKER_URL: 'js/compiled/worker/stellerator_worker.js',
                                VIDEO_PIPELINE_WORKER_URL: 'js/compiled/worker/video_pipeline_worker.js',
                                HELPPAGE_URL: 'doc/stellerator.md',
                                BUILD_ID: buildId
                            }),
                            { global: true }
                        ),
                    browserifyOptions: {
                        debug: true
                    }
                }
            },
            stellerator_prod: {
                dest: 'web/js/compiled/stellerator.prod.js',
                src: ['src/web/stella/stellerator/main.tsx'],
                options: {
                    configure: b =>
                        b.plugin('tsify', { project: __dirname }).transform(
                            envify({
                                NODE_ENV: 'production',
                                STELLERATOR_WORKER_URL: 'js/worker/stellerator_worker.min.js',
                                VIDEO_PIPELINE_WORKER_URL: 'js/worker/video_pipeline_worker.min.js',
                                HELPPAGE_URL: 'doc/stellerator.md',
                                BUILD_ID: buildId
                            }),
                            { global: true }
                        ),
                    browserifyOptions: {
                        debug: true
                    }
                }
            },
            stellerator_embedded: {
                dest: 'web/js/compiled/stellerator_embedded.js',
                src: ['src/web/embedded/stellerator/index.ts'],
                options: {
                    browserifyOptions: {
                        standalone: '$6502',
                        debug: true
                    }
                }
            }
        },

        exorcise: {
            stellerator: {
                src: 'web/js/compiled/stellerator.js',
                dest: 'web/js/compiled/stellerator.js.map'
            },
            stellerator_worker: {
                src: 'web/js/compiled/worker/stellerator_worker.js',
                dest: 'web/js/compiled/worker/stellerator_worker.js.map'
            },
            video_pipeline_worker: {
                src: 'web/js/compiled/worker/video_pipeline_worker.js',
                dest: 'web/js/compiled/worker/video_pipeline_worker.js.map'
            },
            ehBasicCLI: {
                src: 'web/js/compiled/ehBasicCLI.js',
                dest: 'web/js/compiled/ehBasicCLI.js.map'
            },
            debuggerCLI: {
                src: 'web/js/compiled/debuggerCLI.js',
                dest: 'web/js/compiled/debuggerCLI.js.map'
            },
            stellaCLI: {
                src: 'web/js/compiled/stellaCLI.js',
                dest: 'web/js/compiled/stellaCLI.js.map'
            },
            stellerator_embedded: {
                src: 'web/js/compiled/stellerator_embedded.js',
                dest: 'web/js/compiled/stellerator_embedded.js.map'
            }
        },

        uglify: {
            options: {
                sourceMap: true,
                banner: LICENSE_PREAMPLE
            },
            stellerator: {
                dest: 'build/stellerator/js/app.js',
                src: [
                    'web/js/installed/jquery.min.js',
                    'web/js/bootstrap.min.js',
                    'web/js/compiled/stellerator.prod.js'
                ]
            },
            video_pipeline_worker: {
                src: 'web/js/compiled/worker/video_pipeline_worker.js',
                dest: 'web/js/compiled/worker/video_pipeline_worker.min.js',
                options: {
                    sourceMapIn: 'web/js/compiled/worker/video_pipeline_worker.js.map'
                }
            },
            stellerator_worker: {
                src: 'web/js/compiled/worker/stellerator_worker.js',
                dest: 'web/js/compiled/worker/stellerator_worker.min.js',
                options: {
                    sourceMapIn: 'web/js/compiled/worker/stellerator_worker.js.map'
                }
            },
            stellerator_embedded: {
                src: 'web/js/compiled/stellerator_embedded.js',
                dest: 'web/js/compiled/stellerator_embedded.min.js',
                options: {
                    sourceMapIn: 'web/js/compiled/stellerator_embedded.js.map'
                }
            }
        },

        postcss: {
            options: {
                map: false,
                processors: [require('postcss-import')(), require('cssnano')()]
            },
            stellerator: {
                dest: 'build/stellerator/css/app.css',
                src: 'web/css/stellerator.css'
            }
        },

        mochaTest: {
            test: {
                options: {
                    reporter: 'nyan',
                    ui: 'tdd',
                    bail: false,
                    require: 'source-map-support/register'
                },
                src: ['compiled/tests/ts/**/*.js']
            },
            debug: {
                options: {
                    reporter: 'spec',
                    ui: 'tdd',
                    bail: true
                },
                src: ['compiled/tests/ts/**/*.js']
            }
        },

        'http-server': {
            dev: {
                root: 'web',
                port: 6502,
                host: '0.0.0.0',
                autoIndex: true,
                ext: 'html',
                cache: 0,
                runInBackground: true
            },
            build: {
                root: 'build',
                port: 2600,
                host: '0.0.0.0',
                autoIndex: true,
                ext: 'html',
                cache: 0
            }
        },

        blobify: {
            default: {
                options: {
                    baseDir: './aux'
                },
                src: 'aux/**',
                dest: 'web/js/compiled/files.json'
            },
            tests: {
                options: {
                    baseDir: './tests/fixtures/fs_provider',
                    recurse: true
                },
                src: './tests/fixtures/fs_provider/tree',
                dest: './compiled/tests/fixtures/fs_provider/blob.json'
            }
        },

        template: {
            stellerator_dev: {
                src: 'web/template/stellerator.html',
                dest: 'web/stellerator.html',
                options: {
                    data: {
                        stylesheets: ['css/stellerator.css'],
                        scripts: ['js/installed/jquery.min.js', 'js/bootstrap.min.js', 'js/compiled/stellerator.js'],
                        dev: true
                    }
                }
            },
            stellerator_build: {
                src: 'web/template/stellerator.html',
                dest: 'build/stellerator/index.html',
                options: {
                    data: {
                        stylesheets: ['css/app.css'],
                        scripts: ['js/app.js'],
                        dev: false
                    }
                }
            },
            stellerator_appcache: {
                src: 'web/template/stellerator.appcache',
                dest: 'build/stellerator/stellerator.appcache'
            }
        },

        compress: {
            stellerator_embedded: {
                options: {
                    archive: 'dist/stellerator_embedded.zip',
                    mode: 'zip'
                },
                files: [
                    {
                        cwd: 'web/js/compiled',
                        expand: true,
                        src: ['stellerator_embedded.min.js', 'stellerator_embedded.min.js.map']
                    },
                    {
                        cwd: 'web/js/compiled/worker',
                        expand: true,
                        src: ['stellerator_worker.min.js', 'stellerator_worker.min.js.map']
                    }
                ]
            }
        },

        copy: {
            stellerator_prod: {
                files: [
                    {
                        expand: true,
                        cwd: 'web',
                        src: ['css/fonts/**/*'],
                        dest: 'build/stellerator'
                    },
                    {
                        expand: true,
                        src: ['doc/**/*'],
                        dest: 'build/stellerator'
                    },
                    {
                        src: 'web/js/compiled/worker/stellerator_worker.min.js',
                        dest: 'build/stellerator/js/worker/stellerator_worker.min.js'
                    },
                    {
                        src: 'web/js/compiled/worker/video_pipeline_worker.min.js',
                        dest: 'build/stellerator/js/worker/video_pipeline_worker.min.js'
                    },
                    {
                        src: 'web/stellerator_homescreen_icon.png',
                        dest: 'build/stellerator/stellerator_homescreen_icon.png'
                    }
                ]
            },
            stellerator_dev: {
                files: [
                    {
                        expand: true,
                        src: ['doc/**/*'],
                        dest: 'web'
                    }
                ]
            },
            test_fixtures: {
                files: [
                    {
                        expand: true,
                        cwd: 'tests',
                        src: ['fixtures/**/*'],
                        dest: 'compiled/tests'
                    }
                ]
            },
            install: {
                files: {
                    'web/js/installed/jquery.min.js': 'node_modules/jquery/dist/jquery.min.js',
                    'web/js/installed/jquery.terminal.min.js': 'node_modules/jquery.terminal/js/jquery.terminal.min.js',
                    'web/js/installed/jquery.mousewheel.min.js':
                        'node_modules/jquery.terminal/js/jquery.mousewheel-min.js',
                    'web/css/installed/jquery.terminal.min.css': 'node_modules/jquery.terminal/css/jquery.terminal.css'
                }
            }
        },

        typedoc: {
            stellerator_embedded: {
                options: {
                    module: 'commonjs',
                    out: 'typedoc/stellerator-embedded',
                    name: 'Stellerator embedded',
                    target: 'es5',
                    theme: 'minimal',
                    excludePrivate: true,
                    excludeExternals: true,
                    mode: 'modules',
                    readme: 'doc/stellerator_embedded.md',
                    exclude: '**/+(index.ts|ControlPanelProxy.ts|SwitchProxy.ts)'
                },
                src: ['src/web/embedded/stellerator']
            }
        },

        watch: {
            build: {
                files: ['src/**/*.ts', 'aux/**'],
                tasks: ['build']
            }
        },

        notify_hooks: {
            options: {
                enabled: true,
                success: true
            }
        }
    });

    grunt.registerTask('browserify_dev', [
        'browserify:debuggerCLI',
        'browserify:ehBasicCLI',
        'browserify:stellaCLI',
        'browserify:stellerator_dev',
        'browserify:stellerator_worker',
        'browserify:video_pipeline_worker',
        'browserify:stellerator_embedded',
        'exorcise'
    ]);
    grunt.registerTask('browserify_prod', ['browserify:stellerator_prod']);

    grunt.registerTask('template_dev', ['template:stellerator_dev']);

    grunt.registerTask('stellerator:dev', [
        'tslint',
        'browserify:stellerator_dev',
        'browserify:stellerator_worker',
        'browserify:video_pipeline_worker',
        'exorcise:stellerator',
        'exorcise:stellerator_worker',
        'exorcise:video_pipeline_worker',
        'template:stellerator_dev',
        'copy:stellerator_dev'
    ]);

    grunt.registerTask('stellerator:build', [
        'tslint',
        'browserify:stellerator_prod',
        'browserify:stellerator_worker',
        'browserify:video_pipeline_worker',
        'exorcise:stellerator_worker',
        'exorcise:video_pipeline_worker',
        'uglify:stellerator',
        'uglify:stellerator_worker',
        'uglify:video_pipeline_worker',
        'postcss:stellerator',
        'copy:stellerator_prod',
        'template:stellerator_build',
        'template:stellerator_appcache'
    ]);

    grunt.registerTask('stellerator_embedded:build', [
        'tslint',
        'browserify:stellerator_embedded',
        'exorcise:stellerator_embedded',
        'uglify:stellerator_embedded',
        'browserify:stellerator_worker',
        'exorcise:stellerator_worker',
        'uglify:stellerator_worker',
        'compress:stellerator_embedded'
    ]);

    grunt.registerTask('tslint', ['exec:tslint']);
    grunt.registerTask('dev', ['tslint', 'browserify_dev', 'template_dev', 'blobify:default', 'copy:stellerator_dev']);
    grunt.registerTask('build', ['stellerator:build', 'stellerator_embedded:build']);
    grunt.registerTask('test', ['tslint', 'ts:main', 'copy:test_fixtures', 'blobify:tests', 'mochaTest:test']);
    grunt.registerTask('test:debug', ['tslint', 'ts:main', 'copy:test_fixtures', 'blobify:tests', 'mochaTest:debug']);
    grunt.registerTask('initial', ['clean', 'copy:install', 'test']);

    grunt.registerTask('cleanup', ['clean:base', 'clean:worker']);
    grunt.registerTask('mrproper', ['clean']);

    grunt.registerTask('serve', ['http-server:dev', 'http-server:build']);

    grunt.registerTask('default', ['dev']);

    grunt.task.run('notify_hooks');
};
